lib/Service/ImageService.php

```
@@ -0,0 +1,658 @@
+					$fileName = $fileName . '.bmp';
+				} elseif ($res['Content-Type'] === 'image/svg+xml') {
+					$fileName = $fileName . '.svg';
+				} else {
+					$ext = preg_replace('/^image\//i', '', $res['Content-Type']);
+					$fileName = $fileName . '.' . $ext;
+				}
+				$targetPath = $saveDir->getPath() . '/' . $fileName;
+				$savedFile->move($targetPath);
+				// get file type and name
+				return [
+					'name' => $fileName,
+					'id' => $savedFile->getId(),
+					'documentId' => $textFile->getId(),
+				];
+			} else {
+				$savedFile->delete();
+				return [
+					'error' => 'Unsupported file type',
+				];
+			}
+		} elseif (isset($res['error'])) {
+			$savedFile->delete();
+			return $res;
+		} else {
+			$savedFile->delete();
+			return [
+				'error' => 'Link download error',
+			];
+		}
+	}
+
+	/**
+	 * Get or create file-specific attachment folder
+	 *
+	 * @param File $textFile
+	 * @param bool $create
+	 * @return Folder|null
+	 * @throws NotFoundException
+	 * @throws NotPermittedException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OC\User\NoUserException
+	 */
+	private function getAttachmentDirectoryForFile(File $textFile, bool $create = false): ?Folder {
+		$owner = $textFile->getOwner();
+		$ownerId = $owner->getUID();
+		$ownerUserFolder = $this->rootFolder->getUserFolder($ownerId);
+		$ownerTextFile = $ownerUserFolder->getById($textFile->getId());
+		if (count($ownerTextFile) > 0) {
+			$ownerTextFile = $ownerTextFile[0];
+			$ownerParentFolder = $ownerTextFile->getParent();
+			$attachmentFolderName = '.attachments.' . $textFile->getId();
+			if ($ownerParentFolder->nodeExists($attachmentFolderName)) {
+				$attachmentFolder = $ownerParentFolder->get($attachmentFolderName);
+				if ($attachmentFolder instanceof Folder) {
+					return $attachmentFolder;
+				}
+			} elseif ($create) {
+				return $ownerParentFolder->newFolder($attachmentFolderName);
+			}
+		}
+		return null;
```

I wonder whether it would be cleaner to throw a `NotFoundException` instead of returning null in low-level functions like this one, `getFileFromPath()` and `getImagePreview()`. That way you don't have to check for the returned value being null in the higher level functions and don't have to care about null pointer exceptions.

---

Much better, simpler and more readable.
Exceptions are now caught in the controller (except for the delete/copy/move methods in the service).
Done in c204d904ab3a1bba6c58d32d8a615ebab04808b4